draft-ietf-tls-tls13.md

```diff
@@ -2469,24 +2469,40 @@ selected_identity
 
 Each PSK is associated with a single Hash algorithm. For PSKs established
 via the ticket mechanism ({{NewSessionTicket}}), this is the Hash used for
-the KDF. For externally established PSKs, the Hash algorithm MUST be set when the
-PSK is established.
-
-Prior to accepting PSK key establishment, the server MUST validate the
-corresponding binder value (see {{psk-binder}} below). If this value is
-not present or does not validate, the server MUST abort the handshake.
-Servers SHOULD NOT attempt to validate multiple binders; rather they
-SHOULD select a single PSK and validate solely the binder that
-corresponds to that PSK. In order to accept PSK key establishment, the
-server sends a "pre_shared_key" extension indicating the selected
-identity.
-
-Clients MUST verify that the server's selected_identity is within the
-range supplied by the client, that the server selected a cipher suite
-containing a Hash associated with the PSK, and that the "key_share", and
-"signature_algorithms" extensions are consistent with the indicated
-ke_modes and auth_modes values. If these values are not consistent,
-the client MUST abort the handshake with an "illegal_parameter" alert.
+the KDF. For externally established PSKs, the Hash algorithm MUST be
+decided during the establishment of the PSK.
+
+In order to attempt a PSK key establishment, Clients SHOULD advertise at least one
+cipher suite containing a Hash associated with the PSK.
+
+In order to perform a PSK key establishment, Servers SHOULD go through these steps:
+
+1. Servers SHOULD first pick a cipher suite.
+
+2. Servers SHOULD find a PSK Identity matching the Hash of the chosen cipher suite.
+
+3. Servers MUST attempt to validate the Binder associated to the chosen Identity.
+
+If no PSK Identity matching the chosen cipher suite is found, a Server SHOULD
+process with a normal handshake and ignore any other PSK extensions.
+
+If the associated binder value (see {{psk-binder}} below) is not present or does
+not validate, the Server MUST abort the handshake. Servers SHOULD NOT attempt
+to validate multiple binders.
+
+If the binder validates, to finally accept the PSK key establishment, Servers MUST
+send a "pre_shared_key" extension indicating the selected identity.
+
+Note that this process is more involved if the Client wishes to use the early_data
+extension. Not only the Hash function should match, but the entire cipher suite as well.
+Other extensions can also define an early_data requirement to match.
```

Comment:
I'm not sure we really need to be prescriptive here. There's at least one other algorithm:

1. Select PSK if possible.
2. Check to see if the PSK can be used with any client-offered cipher suite and if so do PSK. If not, go do regular negotiation.

This isn't what NSS or BoringSSL does, but I'm not aware of any strong argument for why it's a sufficiently bad algorithm that we should be telling people not to use it.